## <center>Ordered Sets in Data Analysis</center>
<center>Amantur Amatov</center> 
<center>Higher School of Economics</center> 
<center>Moscow, the Russian Federation</center> 
    
----------------------------------------------------

### Task 1.
Let $U$ – finite set, and $f: U\rightarrow U$  
Prove, if $f$ is surjective functions, then $f$ is injective function

### Proof:

Suppose $f$ is not injective. Then $\exists x_1 \neq x_2 \Leftrightarrow f(x_1)=f(x_2)$.  
Then, since $U$ is finite, suppose that $f(x_i)=y_i$ for each $1 \leq i \leq |S|$. So if $f$ is not injective, let $f(x_1)=f(x_2)$, whence $y_1=y_2$, so the set

$$f(U) = \{y_1,y_2,...,y_{|U|}\}$$ 
contains at most $|U|-1$ distinct elements.

So,
$$|f(U)|\leq |U|-1$$

Also, since $f$ is surjective, $f(U)=U$, then 

$$|f(U)|=|U|$$

Then,

$$|U|\leq |U|-1$$ is a contradiction. And then $f$ is **injective**.

### Task 2
Let consider binary relations on the set of 5 elements.  
Count how many different binary relations that are simultaneously asymmetric and transitive

### Solution

By definition:
An asymmetric relation is a binary relation on a set $X$ where for all $a$ and $b$ in $X$, if $a$ is related to $b$, then $b$ is not related to $a$.
In the notation of first-order logic:

$$
\forall a,b \in X: aRb \rightarrow \neg(bRa)
$$

A transitive relation is a relation on a set $X$ where for all $a,b,c \in X$, if $aRb$ and $bRc$, then $aRc$.
In the notation of first-order logic:

$$
\forall a,b,c \in X: (aRb \wedge bRc) \rightarrow aRc
$$

So we have set of 5 elements. Matrix of this set will look like this:

||$x_1$| $x_2$ | $x_3$ | $x_4$ | $x_5$ |
| --- | --- | --- | --- | --- | --- |
|  $x_1$ |&nbsp;| &nbsp; | &nbsp; | &nbsp; | &nbsp; |
|  $x_2$ |&nbsp;| &nbsp; | &nbsp; | &nbsp; | &nbsp; |
|  $x_3$ |&nbsp;| &nbsp; | &nbsp; | &nbsp; | &nbsp; |
|  $x_4$ |&nbsp;| &nbsp; | &nbsp; | &nbsp; | &nbsp; |
|  $x_5$ |&nbsp;| &nbsp; | &nbsp; | &nbsp; | &nbsp; |

So we need to add 21 relations to our number of relations, that we will compute with Python, because, there is 20 relations with only one pair, like $x_1 R x_2$ or $x_2 R x_1$ etc. And these relations are asymmetric and transitive.
And one relation that is an empty set, that is also a relation, that corresponds to conditions.  
So we have **21+N** relations and we need to check other relations with pairs from 3 to 10.  

For solving this task there was written a program in Python 3.

In [ ]:
from itertools import combinations

def generator(n,k):
    result = []
    for bits in combinations(range(n),k):
        s = [0] * n
        for bit in bits:
            s[bit] = 1
        result.append(s)
    return result

In [16]:
def check_relation(R):
    size = R.shape[0]
    for i in range(size):
        if R[i,i]==1:
            return False
        for j in range(size):
            #check for asymmetry
            if R[i,j]==1 and R[j,i]==1:
                return False
            for k in range(size):
                if j==k:
                    continue
                #check for transitivity
                if R[i,j]==1 and R[j,k]==1 and R[i,k]==0:
                    return False
    return True

In [52]:
import numpy as np

N = 0
for i in range(2,11):
    for element in generator(25,i):
        r = np.array(element)
        r = np.reshape(r,(5,5))
        if check_relation(r):
            N+=1
N += 21
print('Relations: {}'.format(N))

Relations: 4231


Answer: **4231**  
We can check it at https://oeis.org/A001035

### Task 3
In this task we ask you to implement topological sort algorithm in any language of your choice (except esoteric ones).  
Compare the computational complexity for 2 cases of input data:
1. List of edges
2. Adjacency matrix

### Solution
In this task topological sort was implemented in Python.  

Realization for Topologic Sort with input data as List of edges:

In [111]:
from collections import defaultdict 

class Graph_edges: 
    def __init__(self, vertices): 
        self.graph = defaultdict(list)
        self.V = vertices
    def addEdge(self, u, v): 
        self.graph[u].append(v) 
  
    def topologicalSort(self): 
        in_degree = [0]*(self.V) 
        for i in self.graph: 
            for j in self.graph[i]: 
                in_degree[j] += 1
        queue = [] 
        for i in range(self.V): 
            if in_degree[i] == 0: 
                queue.append(i) 
        cnt = 0
        top_order = [] 
        while queue: 
            u = queue.pop(0) 
            top_order.append(u) 
            for i in self.graph[u]: 
                in_degree[i] -= 1 
                if in_degree[i] == 0: 
                    queue.append(i) 
            cnt += 1
        return top_order

In [112]:
g = Graph_edges(4) 
g.addEdge(2, 3) 
g.addEdge(3, 0) 
g.addEdge(3, 1) 
g.addEdge(2, 0)
g.topologicalSort()

[2, 3, 0, 1]

If a directed acyclic graph with the set  $V$  of vertices and the set $E$ of edges is represented as list of edge, then complexity of algorithm is $O(V+E)$. The traverse of the edges to construct a list of in-degrees for vertices will take $|E|$ steps and the traverse the edges to find all edges that connect each vertice to other vertices will take $|V|$ steps.

Realization for Topologic Sort with input data as Adjacency matrix:

In [113]:
class Graph_matrix:
    def __init__(self, adj_matrix):
        self.graph = adj_matrix
        self.count = range(len(self.graph))

    def TopologicalSort(self):
        top_order =[]
        workArray = []
        for i in self.count: 
            step=0
            for j in self.count:
                if self.graph[i][j]==1:step+=1
            workArray.insert(i,step)
        completedCounter = 0
        currentLevel = 0
        while (completedCounter != len(self.graph)):
            for i in self.count:
                if (workArray[i] == 0):
                    ind=0
                    top_order.insert(completedCounter,i)
                    for node in self.graph:
                        if node[i]==1:
                            workArray[ind]-=1
                        ind+=1

                    workArray[i] = -1
                    completedCounter+=1
            currentLevel+=1
        return top_order

In [114]:
#edges
#(2, 3) 
#(3, 0) 
#(3, 1) 
#(2, 0)
g = Graph_matrix([[0,0,1,1],
                  [0,0,0,1],
                  [0,0,0,0],
                  [0,0,1,0]])
g.TopologicalSort()

[2, 3, 0, 1]

If a directed acyclic graph with the set $V$ of vertices and the set E of edges is represented as adjecency matrix, so then realization would require $|V|^2$ steps,because we define every relation between edges. So the whole algorithm will execute in $O(|V|^2)$

### Task 4

Let us consider a binary relation on $Z^2$:
$$ R(x_1,y_1)R(x_2,y_2)\leftrightarrow x_1\leq x_2, y_1\leq y_2$$
Find minimal and maximal elements if R is defined on following sets:
1. $A_1 = \{(x,y)|x\leq 3, y\leq 4\}$
2. $A_2 = \{(x,y)|x^2+y^2 \leq 4\}$

### Solution
Let's check if $R$ is a partial order on $\mathbb{Z}^2$:

$1.$ reflexivity $\forall x: xRx$ :
$$ \begin{cases}
               x \leq x \\
               y \leq y
            \end{cases}  \Rightarrow  (x,y)R(x,y)
$$
            
$2.$ antisymmetricity: $\forall x,y: xRy \wedge yRx \Rightarrow x = y $ :
$$  \begin{cases}
               (x_1, y_1)R(x_2, y_2) \\
               (x_2, y_2)R(x_1, y_1)
            \end{cases} \iff \begin{cases}
               x_1 \leq x_2 \leq x_1 \\
               y_1 \leq y_2 \leq y_1
            \end{cases} \iff \begin{cases}
               x_1 = x_2  \\
               y_1 = y_2 
            \end{cases} \iff (x_1, y_1) = (x_1, y_2)
$$

$3.$ transitivity:  $\forall x,y,z: xRy \wedge yRz \Rightarrow xRz $

$$  \begin{cases}
               (x_1, y_1)R(x_2, y_2) \\
               (x_2, y_2)R(x_3, y_3)
            \end{cases} \iff \begin{cases}
               x_1 \leq x_2 \leq x_3 \\
               y_1 \leq y_2 \leq y_3
            \end{cases} \iff \begin{cases}
               x_1 \leq x_3  \\
               y_1 \leq y_3 
            \end{cases} \iff (x_1, y_1)R(x_3, y_3)
$$

Let's find minimal and maximal element 

$(3, 4)$ is obviously maximal element of $A_1$ (by definition partial order $R$). 

There are no minimal element in the set. Assume, $a = (x_1, y_1)$ is the minimal element in $A_1$. But there are $(x_1-1, y_1)Ra$, so a is not the minimal element

$ A_2 = {(−2,0), (−1, −1), (−1,0), (−1,1), (0, −2), (0, −1), (0,0), (0,1), (0,2), (1, −1), (1,0), (1,1), (2,0)} $
Maximal elements are ${(0,2), (1,1), (2,0)}$
Minimal elements are ${(0,-2), (-1,-1), (-2,0)}$

For $A_1$ maximum is $(3, 4)$, no minimal elements exists.

For $A_2$ maximum elements are ${(0,2), (1,1), (2,0)}$, minimals elements are ${(0,-2), (-1,-1), (-2,0)}$.

### Task 5
Which of the following figures represent a lattice? Why?  
Are there any complete lattices?
![alt text](lattice.png)

### Solution
By definition of lattice it consists of a partially ordered set in which every two elements have a **unique** supremum and a **unique** infimum.
1) **No**: no supremum for a and b, because ${c,d,e}$ are supremums for $a$ and $b$, but $c$ and $d$ are incomparable, and so for other pairs of elements.
![alt text](1lattice.png)  
2) **Yes**: every two elements have a **unique** supremum and a **unique** infimum.  
3) **Yes**: every two elements have a **unique** supremum and a **unique** infimum.  
4) **Yes**: every two elements have a **unique** supremum and a **unique** infimum.  
5) **No**:  no supremum for two elements ${a,b}$
![alt text](5lattice.png)

Lattices 2,3,4 are complete lattices because of their finitness. 

### Task 6
Let $(L, \leq)$ be a lattice with infimum and sumpremum defined as usual.
Prove that for any $x,y \in L$
1. $x \vee x = x$
2. $x \vee (x \wedge y)=x$
3. $x \vee y = y \vee x$

### Solution
1. By definition, $x \vee x = sup(x)=\hat{x} \in L$ such that $x \leq \hat{x}$ and $\forall l \in L x \leq l$ implies that $\hat{x} \leq l$. Since $L$ is partially ordered, we know that $x \leq x$, but by definition of supremum $\hat{x} \leq x$. But we assumed that $x \leq \hat{x}$ is antisymmetric, it follows that $x = \hat{x}=x\vee x$
2. $x \vee (x \wedge y)=sup(x,inf(x,y))=sup(x,a)=b$, where 
    - $a \leq x, a \leq y$
    - $b \geq x, b \geq a$  
Suppose, that $b = x$. 
Then, $x \geq x, x \geq a$. It satisfies the first part.  
And suppose there is another $\hat{b}$ that $\hat{b} \geq x, \hat{b} \geq a$. Then $x \geq \hat{b}$. Due to antisymmetry,
$$((x \geq \hat{b}) \wedge (\hat{b} \geq x)) \rightarrow (\hat{b}=x) \rightarrow  x=sup(x,a)=sup(x,inf(x,y))= x \vee (x \wedge y)$$
3. By definition $x \vee y = a$ and $y \vee x = b$ and $a \geq x$,$a\geq y$. $b \geq x$ and $b\geq y$. Let's pretend that $a>b$, it contradicts the definition of supremum of two elements in lattice. So $b>a$ is. That means, that $a=b \rightarrow x \vee y = y \vee x$ 